Move logging to a callback.

iphydf opened the pull request:
This removes the global logger (which by the way was deleted when the first tox
was killed, so other toxes would then stop logging). Various bits of the code
now carry a logger or pass it around. It's a bit less transparent now, but now
there is no need to have a global logger, and clients can decide what to log and
where.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/toxcore/46)
<!-- Reviewable:end -->

iphydf commented:
@linux-modder you were talking about community reviews. This change is a good one to test your process on.

mannol commented:
OK. First time using reviewable... Anyways, seems like it will require more effort to write them log messages but an overall a positive change.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KP_I9WaG7KBPjVQnQ4V:bilp1v2)*
<!-- Sent from Reviewable.io -->

tux3 commented on toxdns/toxdns.c:
```diff
@@ -144,7 +144,8 @@ int tox_generate_dns3_string(void *dns3_object, uint8_t *string, uint16_t string
 
     if (end_len != string - old_str) {
-        LOGGER_ERROR("tox_generate_dns3_string Fail, %u != %lu\n", end_len, string - old_str);
+        // This currently has no access to a logger.
```
It would be nice to add a "FIXME:" for easier grepability

iphydf replied:
Done.

tux3 commented on toxcore/network.c:
```diff
@@ -265,29 +265,35 @@ uint64_t current_time_monotonic(void)
+    return buflen > 7 ? ntohl(*(uint32_t *)&buffer[5]) : 0;
+}
+
+static void loglogdata(Logger *log, const char *message, const uint8_t *buffer,
```
This function can be called very frequently, and all the LOGGER_TRACE calls with their arguments will add a non-negligible overhead on weaker systems. 
Since the TRACE log level will rarely be helpful to users or client devs checking the log, it would be nice to have a compile-time define for the miminum log level.

For example if I pass -DLOG_LEVEL=WARNING to the compiler, I'd like TRACE, DEBUG and INFO logger calls to be compiled out entirely, with something reasonnable like INFO being the default.

Would it be possible to add something like that?

iphydf replied:
Done.

tux3 replied:
Thanks, looks good to me.

tux3 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 38 of 38 files at r1, 6 of 6 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KPd8MvTGOWFCF1cFWH2:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: 31 of 37 files reviewed at latest revision, 5 unresolved discussions.

---

*[toxav/bwcontroller.c, line 158 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPePlS9zRgi1QQGf-eK:-KPePlS9zRgi1QQGf-eL:bobq3gj) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxav/bwcontroller.c#L158)):*
> ```C
>         if (bwc->cycle.lost) {
>             LOGGER_DEBUG(bwc->m->log, "%p Sent update rcv: %u lost: %u",
>                          bwc, bwc->cycle.recv, bwc->cycle.lost);
> ```

Why is messenger's logger used in here? Does this function run in messenger's thread? Is the log callback in `tox.h` thread-safe? Does the `tox.h` log callback get called only while inside `tox_iterate()`?

---

*[toxcore/logger.c, line 67 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPeXt0ApyEcoY4iUYmq:-KPeXt0ApyEcoY4iUYmr:bz2etkv) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/logger.c#L67)):*
Why limit debug message to 1024 characters? We could make `msg` arbitrarily long. It would have to be allocated on heap though, which would result in rather big performance penalty if called very often. To mitigate the performance overhead, we could keep the 1024 char `msg` on the stack and `vsnprintf()` into it, but if `vsnprintf()` tells us that these 1024 bytes weren't enough, only then allocate a buffer on heap and `vsnprintf()` into it? `vnsprintf()` returns 

>The number of characters that would have been written if n had been sufficiently large, not counting the terminating null character.

so we would know exactly how much space to allcoate on the heap.

---

*[toxcore/tox.c, line 133 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPebHCNAX_5bSRQThGj:-KPebHCNAX_5bSRQThGk:b2gwsr6) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/tox.c#L133)):*
> ```C
>     _Bool load_savedata_sk = 0, load_savedata_tox = 0;
> 
>     Logger *log = logger_new(random_int());
> ```

Should it really be random? We could keep a global counter ourselves, or let the user pass an integer to `tox_new()`. Making the user pick it makes sense, that way user would know which Tox instance the Id corresponds to.

Actually, do we even need that integer? What is the purpose of it? A user would be able to know which Tox instance is calling the log callback because the callback would have the `userdata` argument that the user has passed to that instance's `tox_iterate()', so I don't see a point in that integer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 38 of 38 files at r1.
Review status: 31 of 37 files reviewed at latest revision, 5 unresolved discussions.

---

*[toxav/bwcontroller.c, line 158 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPePlS9zRgi1QQGf-eK:-KPefdKZSVraczk2UyW3:bc67lr5) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxav/bwcontroller.c#L158)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Why is messenger's logger used in here? Does this function run in messenger's thread? Is the log callback in `tox.h` thread-safe? Does the `tox.h` log callback get called only while inside `tox_iterate()`?
</details>
Hm, I see there is really no toxav's logger, it's the same logger used in messenger and av. Doesn't that result in any threading issues as far as the log callback from the `tox.h` is called?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 6 of 6 files at r2, 3 of 3 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[toxav/bwcontroller.c, line 158 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPePlS9zRgi1QQGf-eK:-KPlGY1Sqli0iNfljlWR:bw0oo5i) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxav/bwcontroller.c#L158)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Hm, I see there is really no toxav's logger, it's the same logger used in messenger and av. Doesn't that result in any threading issues as far as the log callback from the `tox.h` is called?
</details>
Yes, it does. I've added a comment about this to the callback docs.

---

*[toxcore/logger.c, line 67 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPeXt0ApyEcoY4iUYmq:-KPlGeMl-HpcNNofFXGc:b-r89237) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/logger.c#L67)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Why limit debug message to 1024 characters? We could make `msg` arbitrarily long. It would have to be allocated on heap though, which would result in rather big performance penalty if called very often. To mitigate the performance overhead, we could keep the 1024 char `msg` on the stack and `vsnprintf()` into it, but if `vsnprintf()` tells us that these 1024 bytes weren't enough, only then allocate a buffer on heap and `vsnprintf()` into it? `vnsprintf()` returns 
> 
> >The number of characters that would have been written if n had been sufficiently large, not counting the terminating null character.
> 
> so we would know exactly how much space to allcoate on the heap.
</details>
The limit used to be 4096. I could make it 4096 again, but it seemed a bit pointless to support huge log messages. I'd like to keep it this way and reconsider if someone comes up with a good reason to write huge log messages. Right now, none of the logger calls write large messages.

---

*[toxcore/tox.c, line 133 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPebHCNAX_5bSRQThGj:-KPlH5FxokgFfi1fMj6Y:b-hnypmp) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/tox.c#L133)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Should it really be random? We could keep a global counter ourselves, or let the user pass an integer to `tox_new()`. Making the user pick it makes sense, that way user would know which Tox instance the Id corresponds to.
> 
> Actually, do we even need that integer? What is the purpose of it? A user would be able to know which Tox instance is calling the log callback because the callback would have the `userdata` argument that the user has passed to that instance's `tox_iterate()', so I don't see a point in that integer.
</details>
Removed the id.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[toxcore/logger.c, line 67 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPeXt0ApyEcoY4iUYmq:-KPqYzAeh0E3Dii_nPWV:b2nhll9) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/logger.c#L67)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> The limit used to be 4096. I could make it 4096 again, but it seemed a bit pointless to support huge log messages. I'd like to keep it this way and reconsider if someone comes up with a good reason to write huge log messages. Right now, none of the logger calls write large messages.
</details>
Ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[toxcore/tox.c, line 133 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPebHCNAX_5bSRQThGj:-KPqabK3lyAOR0_bem42:bhz8ci1) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/tox.c#L133)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Removed the id.
</details>
Didn't realize the log callback in `tox.h` didn't have `userdata` as an argument. Given the threading complications `userdata` imposes, I now get a glimpse of the possible idea you had behind having that integer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[toxcore/tox.c, line 133 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPebHCNAX_5bSRQThGj:-KPqbDIsr3LLBVnK9413:bkephle) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/tox.c#L133)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Didn't realize the log callback in `tox.h` didn't have `userdata` as an argument. Given the threading complications `userdata` imposes, I now get a glimpse of the possible idea you had behind having that integer.
</details>
Oh, it's not the callback that didn't have it but the callback registration function. Ignore my last message. Too bad one can't delete messages from Reviewable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 9 of 9 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/logger.c, line 31 \[r5\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPq_Zoj-gNSzrShNfY0:-KPq_ZokC7ZMaOy092Qe:b-se24ml) ([raw file](https://github.com/toktok/toxcore/blob/d3a153747924dd077b66579e0011eb99c1ec6209/toxcore/logger.c#L31)):*
> ```C
> struct Logger {
>     logger_cb *callback;
>     void *userdata;
> ```

Will this eventually be made stateless too (#40), e.g. in some future PR? It would seem to be hard to make it follow #40 because of possible race conditions when modifying `userdata` inside callbacks or in `tox_iterate()` calls from different threads. Asking the user to use a mutex to guard modification of the `userdata` inside the log callback and guard the `tox_iterate()` call should work, i think (haven't given it too much thought though, so don't take my word).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 38 of 38 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[toxcore/tox.c, line 251 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPrpciU7Vzsrb7pVRqp:-KPrpciVhZwkotSGj_yh:btglieq) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/tox.c#L251)):*
>  Messenger *m = tox;

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KPrpvlpL5SuIzt7p94j:bnfp4nl)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[toxcore/logger.c, line 31 \[r5\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPq_Zoj-gNSzrShNfY0:-KPsyBkK1iZiCs6TjeCy:b4azshc) ([raw file](https://github.com/toktok/toxcore/blob/d3a153747924dd077b66579e0011eb99c1ec6209/toxcore/logger.c#L31)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Will this eventually be made stateless too (#40), e.g. in some future PR? It would seem to be hard to make it follow #40 because of possible race conditions when modifying `userdata` inside callbacks or in `tox_iterate()` calls from different threads. Asking the user to use a mutex to guard modification of the `userdata` inside the log callback and guard the `tox_iterate()` call should work, i think (haven't given it too much thought though, so don't take my word).
</details>
The logging callback might forever be stateful. This might be an acceptable compromise, given that logging could happen in any function. FFI can ignore the logging callback or implement a different kind of abstraction on it or do whatever it takes to expose logging to the target language. The key point is that all other functions are strictly necessary for a fully functional client library, while the logging callback is just for debugging.

If at some point we do really want to make the logging callback stateless, we'll need to temporarily store log messages and ship them to the client in `tox_iterate`. I don't think there is much benefit in doing that right now.

---

*[toxcore/tox.c, line 251 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KPrpciU7Vzsrb7pVRqp:-KPsy7heN2m1ugmNAhPB:b228yug) ([raw file](https://github.com/toktok/toxcore/blob/b868b275a633232aa7b4e073db9c4bb372c23485/toxcore/tox.c#L251)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary>
> >  Messenger *m = tox;
</details>
What does this mean?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Is everybody happy with this PR? Please take another look and comment if you find any potential issues.

mannol commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 38 of 38 files at r1.
Review status: 33 of 37 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KQ7fMidUrMpxKBT4ovx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mannol commented:
Reviewed 1 of 3 files at r3, 1 of 6 files at r4, 8 of 9 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

mannol commented:
[merge it](https://twitter.com/ThePracticalDev/status/769236757065990144)

nurupo commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/logger.h, line 66 \[r6\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KQ8E3WmcaXypguJTODw:-KQ8E3WmcaXypguJTODx:by3thqb) ([raw file](https://github.com/toktok/toxcore/blob/ec091ebe0c5dc740c0955ef3c16f4e4f485e0609/toxcore/logger.h#L66)):*
> ```objc
> #define LOGGER_WRITE(log, level, format, ...) \
>     do { \
>         if (level >= MIN_LOGGER_LEVEL) { \
> ```

`MIN_LOGGER_LEVEL` is a compile-time option, but this check is done at run-time. Any way to make this check compile-time too?

Alternatively, why not make the log level a run-time option, so that a client could select the logging level, since no code seems to be compiled-out anyway and also because with the way this PR is now, the same client would have different logs depending on how the toxcore was built. Also, what about a getter function for the client to know the logging level toxcore uses, would that be useful?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/logger.h, line 66 \[r6\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KQ8E3WmcaXypguJTODw:-KQ8LnwVRsaDdEdhwcef:b-ixv1xj) ([raw file](https://github.com/toktok/toxcore/blob/ec091ebe0c5dc740c0955ef3c16f4e4f485e0609/toxcore/logger.h#L66)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> `MIN_LOGGER_LEVEL` is a compile-time option, but this check is done at run-time. Any way to make this check compile-time too?
> 
> Alternatively, why not make the log level a run-time option, so that a client could select the logging level, since no code seems to be compiled-out anyway and also because with the way this PR is now, the same client would have different logs depending on how the toxcore was built. Also, what about a getter function for the client to know the logging level toxcore uses, would that be useful?
</details>
I made it an option at all because there was a concern about TRACE being too computationally intensive. This is effectively a compile time option now, because compilers are permitted to (and do) remove the compile time constant `if (false)` branch.

I can make it a runtime option. The performance probably wouldn't be degraded a lot, it would just turn those compile time branches into runtime ones.

Opinions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/logger.h, line 66 \[r6\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KQ8E3WmcaXypguJTODw:-KQ8Voj7Y8HR6uDs3_hk:blshonr) ([raw file](https://github.com/toktok/toxcore/blob/ec091ebe0c5dc740c0955ef3c16f4e4f485e0609/toxcore/logger.h#L66)):*
>This is effectively a compile time option now, because compilers are permitted to (and do) remove the compile time constant if (false) branch.

Ok, didn't realize that. Then it's good.

>Opinions?

I don't have a strong opinion about it being a compile-time or run-time.  I guess compile-time is somewhat mobile-friendly, less cpu cycles and battery usage (and tux3 has this dead-slow laptop he likes to run stuff on :P), so I'm more inclined for compile-time. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

nurupo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KQ97_YmG9zh3cyBgQRe:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[toxcore/logger.h, line 67 \[r7\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KQ9wZEnJ6XZTTrlCmDr:-KQ9wZEnJ6XZTTrlCmDs:bsvta8n) ([raw file](https://github.com/toktok/toxcore/blob/13ae9e9a93a1c02fad9475002c0391b86b7ad7bb/toxcore/logger.h#L67)):*
> ```objc
>     do { \
>         if (level >= MIN_LOGGER_LEVEL) { \
>             logger_write(log, level, __FILE__, __LINE__, __func__, format, ##__VA_ARGS__); \
> ```

`##__VA_ARGS__` [is a gcc extension](https://gcc.gnu.org/onlinedocs/gcc/Variadic-Macros.html), do we want to use it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[toxcore/logger.h, line 67 \[r7\]](https://reviewable.io:443/reviews/toktok/toxcore/46#-KQ9wZEnJ6XZTTrlCmDr:-KQAUHNQl0yuhJn-hjCs:bujnaps) ([raw file](https://github.com/toktok/toxcore/blob/13ae9e9a93a1c02fad9475002c0391b86b7ad7bb/toxcore/logger.h#L67)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> `##__VA_ARGS__` [is a gcc extension](https://gcc.gnu.org/onlinedocs/gcc/Variadic-Macros.html), do we want to use it?
</details>
This is not a change. It was done like this before, so I kept it. Fixing it would involve a bit of effort I'd like to avoid in this PR. Feel free to make a followup PR to get rid of the gcc extension.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KQAfrQAONS0Z-u7E_gO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
Reviewed 23 of 38 files at r1, 1 of 3 files at r3, 1 of 6 files at r4, 8 of 9 files at r5, 3 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r3, 1 of 6 files at r4, 8 of 9 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KQBQgCH4SMZ6sbjv26o:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mannol commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/46#-:-KQBRESCiqaPSzw6ME_I:bnfp4nl)*
<!-- Sent from Reviewable.io -->
